client/executor/wasmtime/src/imports.rs

```diff
@@ -0,0 +1,225 @@
+pub fn resolve_imports(
+	state_holder: &StateHolder,
+	module: &Module,
+	host_functions: &[&'static dyn Function],
+) -> Result<Vec<Extern>, WasmError> {
+	let mut externs = vec![];
+	for import_ty in module.imports() {
+		if import_ty.module() != "env" {
+			return Err(WasmError::Other(format!(
+				"host doesn't provide any imports from non-env module: {}:{}",
+				import_ty.module(),
+				import_ty.name()
+			)));
+		}
+
+		let host_func = host_functions
+			.iter()
+			.find(|host_func| host_func.name() == import_ty.name())
+			.ok_or_else(|| {
+				WasmError::Other(format!(
+					"host doesn't provide such function: {}:{}",
+					import_ty.module(),
+					import_ty.name()
+				))
+			})?;
+
+		let func_ty = match import_ty.ty() {
+			ExternType::Func(func_ty) => func_ty,
+			_ => {
+				return Err(WasmError::Other(format!(
+					"host doesn't provide any non function imports: {}:{}",
+					import_ty.module(),
+					import_ty.name()
+				)));
+			}
+		};
+		if !signature_matches(&func_ty, &wasmtime_func_sig(*host_func)) {
+			return Err(WasmError::Other(format!(
+				"signature mismatch for: {}:{}",
+				import_ty.module(),
+				import_ty.name()
+			)));
+		}
+
+		externs.push(create_host_func_handler(
+			*host_func,
+			module.store(),
+			state_holder.clone(),
+		));
+	}
+	Ok(externs)
+}
+
+/// Returns `true` if `lhs` and `rhs` represent the same signature.
+fn signature_matches(lhs: &wasmtime::FuncType, rhs: &wasmtime::FuncType) -> bool {
+	lhs.params() == rhs.params() && lhs.results() == rhs.results()
+}
+
+/// Wraps the given `host_func` as a wasmtime's `Extern` suitable for passing it as an import.
+fn create_host_func_handler(
+	host_func: &'static dyn Function,
+	store: &Store,
+	state_holder: StateHolder,
+) -> Extern {
+	let func_ty = wasmtime_func_sig(host_func);
+	let callable = HostFuncHandler::new(state_holder, host_func);
+	let func = Func::new(store, func_ty, Rc::new(callable));
+	Extern::Func(func)
+}
+
+struct HostFuncHandler {
+	state_holder: StateHolder,
+	host_func: &'static dyn Function,
+}
+
+impl HostFuncHandler {
+	fn new(state_holder: StateHolder, host_func: &'static dyn Function) -> Self {
+		Self {
+			state_holder,
+			host_func,
+		}
+	}
+}
+
+impl Callable for HostFuncHandler {
+	fn call(
+		&self,
+		wasmtime_params: &[Val],
+		wasmtime_results: &mut [Val],
+	) -> Result<(), wasmtime::Trap> {
+		let unwind_result = self.state_holder.with_context(|host_ctx| {
+			let mut host_ctx = host_ctx.expect(
+				"host functions can be called only from wasm instance;
+				wasm instance is always called initializing context;
+				therefore host_ctx cannot be None;
+				qed
+				"
+			);
+			// `into_value` panics if it encounters a value that doesn't fit into the values
+			// available in substrate.
+			//
+			// This, however, cannot happen since the signature of this function is created from
+			// a `dyn Function` signature of which cannot have a non substrate value by definition.
```

Comment:
What if the values that we support change over time? Would the documentation above still be valid?

Reply:
I guess no. But I am not sure what would be a better way?